Code/GraphMol/MolEnumerator/MolEnumerator.h

```diff
@@ -0,0 +1,106 @@
+//
+//  Copyright (C) 2020 Greg Landrum and T5 Informatics GmbH
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+#ifndef RDKIT_MOLENUMERATOR_H
+#define RDKIT_MOLENUMERATOR_H
+
+#include <RDGeneral/export.h>
+#include <GraphMol/RDKitBase.h>
+#include <GraphMol/MolBundle.h>
+
+#include <vector>
+#include <map>
+#include <string>
+#include <memory>
+
+namespace RDKit {
+namespace MolEnumerator {
+
+//! abstract base class for the a molecule enumeration operation
+class RDKIT_MOLENUMERATOR_EXPORT MolEnumeratorOp {
+ public:
+  MolEnumeratorOp(){};
+  virtual ~MolEnumeratorOp(){};
+  //! returns a vector of the number of possible variations at variability point
+  //! covered by this operation
+  virtual std::vector<size_t> getVariationCounts() const = 0;
+  //! returns a the molecule corresponding to a particular variation
+  /*!  which.size() should be equal to the number of variation counts.
+      Note that the caller owns the returned pointer
+  */
+  virtual ROMol *operator()(const std::vector<size_t> &which) const = 0;
+  //! initializes this operation to work on a particular molecule
+  virtual void initFromMol(const ROMol &mol) = 0;
+  //! polymorphic copy
+  virtual MolEnumeratorOp *copy() const = 0;
```

Comment:
I've called this polymorphiic copy "clone" else where in the codebase, we should pick one (copy is fine) and deprecate the others.

Reply:
Good point.
Looks like the `copy()` form is used more frequently, so I will do a PR to rename the `clone()` usages in StreamOps and the FP generator.